src/Compilers/Core/Portable/InternalUtilities/StackGuard.cs

```diff
@@ -29,5 +30,38 @@ public static void EnsureSufficientExecutionStack(int recursionDepth)
                 RuntimeHelpers.EnsureSufficientExecutionStack();
             }
         }
+
+        private static bool TryEnsureSufficientExecutionStack<TArg>(int recursionDepth, Func<TArg, bool> throwOnFailure, TArg arg)
+        {
+            try
+            {
+                EnsureSufficientExecutionStack(recursionDepth);
+                return true;
+            }
+            catch (InsufficientExecutionStackException) when (!throwOnFailure(arg))
+            {
+                return false;
+            }
+        }
+
+        internal static TResult Execute<TResult, TArg1, TArg2>(ref int recursionDepth, Func<TArg1, bool> throwOnFailure, Func<TArg1, TArg2, TResult> execute, TArg1 arg1, TArg2 arg2)
+        {
+            recursionDepth++;
+            TResult result;
+            if (TryEnsureSufficientExecutionStack(recursionDepth, throwOnFailure, arg1))
+            {
+                result = execute(arg1, arg2);
+            }
+            else
+            {
+                var task = Task.Run(() => execute(arg1, arg2));
+                // Wait on the task without inlining the task on this thread.
+                Task.WhenAny(task).Wait();
+                // Return result, propagating any exception.
+                result = task.GetAwaiter().GetResult();
+            }
+            recursionDepth--;
```

Comment:
This won't decrement if the execution throws an exception.. that ok?

Reply:
I don't believe we're recovering from any exceptions within the Parser so it should be ok. But if the overhead of try/finally is low, I'll decrement the value in a finally block to be safe.

---
In reply to: [357835854](https://github.com/dotnet/roslyn/pull/40367#discussion_r357835854) [](ancestors = 357835854)

Reply:
> if the overhead of try/finally is low

If you care about the try/finally overhead, you'll also want to tweak TryEnsureSufficientExecutionStack.  In particular, right now it's not inlineable because it contains exception handling.  If you lift out the recursionDepth check, you could avoid that some of the time, e.g.
```
if (recursionDepth <= MaxUncheckedRecursionDepth && TryEnsureSufficientExecutionStack(throwOnFailure, arg1))
...